### PR TITLE
[CALCITE-844] ProjectWindowTransposeRule added to push LogicalProject…

### PR DESCRIPTION
**Pull request by hsuanyi**
… past LogicalWindow

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/rel/rules/ProjectWindowTransposeRule.java [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+      return;
+    }
+
+    // Put a DrillProjectRel below LogicalWindow
+    final List<RexNode> exps = new ArrayList<>();
+    final List<RelDataTypeField> fields = new ArrayList<>();
+
+    // Keep only the fields which are referred
+    for  (int index = beReferred.nextSetBit(0); index >= 0;
+        index = beReferred.nextSetBit(index + 1)) {
```
Comment (julianhyde) [inline core/src/main/java/org/apache/calcite/rel/rules/ProjectWindowTransposeRule.java]: Use `for(int index : BitSets.toIter(...))`

File: core/src/main/java/org/apache/calcite/rel/rules/ProjectWindowTransposeRule.java [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+      exps.add(new RexInputRef(index, relDataTypeField.getType()));
+      fields.add(relDataTypeField);
+    }
+
+    final LogicalProject projectBelowWindow = new LogicalProject(
+        window.getCluster(),
+        window.getTraitSet(),
+        window.getInput(),
+        exps,
+        new RelRecordType(fields));
```
Comment (julianhyde) [inline core/src/main/java/org/apache/calcite/rel/rules/ProjectWindowTransposeRule.java]: Don't create types directly. Use a typeFactory. Or in this case, use typeFactory.builder().

File: core/src/test/java/org/apache/calcite/test/RelOptRulesTest.java [hunk 3]
```diff
@@ -139,6 +141,15 @@ protected DiffRepository getDiffRepos() {
     return DiffRepository.lookup(RelOptRulesTest.class);
   }
```
Comment (julianhyde) [inline core/src/test/java/org/apache/calcite/test/RelOptRulesTest.java]: don't add tests at the top of the file (or at the bottom). they cause merge conflicts. put them in some "logical" place.

File: core/src/main/java/org/apache/calcite/rel/rules/ProjectWindowTransposeRule.java (hunk 1 above)
Comment (hsuanyi) [inline core/src/main/java/org/apache/calcite/rel/rules/ProjectWindowTransposeRule.java]: Changed line 79 to:   
for  (int index : BitSets.toIter(beReferred)) {

File: core/src/main/java/org/apache/calcite/rel/rules/ProjectWindowTransposeRule.java (hunk 2 above)
Comment (hsuanyi) [inline core/src/main/java/org/apache/calcite/rel/rules/ProjectWindowTransposeRule.java]: Use below to create type:
final RelDataType rowtype = RexUtil.createStructType(
        window.getCluster().getTypeFactory(),
        exps,
        fieldNames);

**Pull request closed without merging**